llpc/context/llpcCompiler.cpp

```diff
@@ -661,24 +661,176 @@ Result Compiler::buildGraphicsShaderStage(const GraphicsPipelineBuildInfo *pipel
   GraphicsContext graphicsContext(m_gfxIp, pipelineInfo, &pipelineHash, &cacheHash);
   Context *context = acquireContext();
   context->attachPipelineContext(&graphicsContext);
+  auto onExit = make_scope_exit([&] { releaseContext(context); });
 
   ElfPackage candidateElf;
   Result result =
       buildUnlinkedShaderInternal(context, shaderInfo, stage, candidateElf, pipelineOut->stageCacheAccesses);
 
-  releaseContext(context);
-
   if (result != Result::Success)
     return result;
 
-  void *allocBuf = pipelineInfo->pfnOutputAlloc(pipelineInfo->pInstance, pipelineInfo->pUserData, candidateElf.size());
+  unsigned metaDataSize = 0;
+  SmallVector<FsOutInfo, 8> fsOuts;
+  bool discardState = false;
+  if (stage == UnlinkedStageFragment && pipelineInfo->options.explicitColorExport) {
+    // Parse ELF to get outputs
+    ElfWriter<Elf64> writer(m_gfxIp);
+    if (writer.ReadFromBuffer(candidateElf.data(), candidateElf.size()) == Result::Success) {
+      ElfNote metaNote = writer.getNote(Abi::MetadataNoteType); // NT_AMDGPU_METADATA
+      msgpack::Document document;
+      document.readFromBlob(StringRef(reinterpret_cast<const char *>(metaNote.data), metaNote.hdr.descSize), false);
+      auto pipeNode =
+          document.getRoot().getMap(true)[PalAbi::CodeObjectMetadataKey::Pipelines].getArray(true)[0].getMap(true);
+
+      // Color export infos
+      auto it = pipeNode.find(Vkgc::ColorExports);
+      if (it != pipeNode.end()) {
+        assert(it->second.isArray());
+        auto colorExports = it->second.getArray();
+        for (unsigned i = 0; i < colorExports.size(); ++i) {
+          msgpack::ArrayDocNode fetchNode = colorExports[i].getArray();
+          FsOutInfo output;
+          output.hwColorTarget = fetchNode[0].getUInt();
+          output.location = fetchNode[1].getUInt();
+          output.isSigned = fetchNode[2].getBool();
+          StringRef typeName = fetchNode[3].getString();
+          memset(output.typeName, 0, sizeof(output.typeName));
+          memcpy(output.typeName, typeName.data(), typeName.size());
+          fsOuts.push_back(output);
+        }
+        metaDataSize = sizeof(FragmentOutputs) + sizeof(FsOutInfo) * fsOuts.size();
+        // Clear
+        pipeNode.erase(it);
+      }
+      // Discard state
+      it = pipeNode.find(Vkgc::DiscardState);
+      if (it != pipeNode.end()) {
+        discardState = it->second.getBool();
+        pipeNode.erase(it);
+      }
+
+      // Overwrite elf
+      ElfNote newMetaNote = metaNote;
+      std::string destBlob;
+      document.writeToBlob(destBlob);
+      auto newData = new uint8_t[destBlob.size()];
+      memcpy(newData, destBlob.data(), destBlob.size());
+      newMetaNote.hdr.descSize = destBlob.size();
+      newMetaNote.data = newData;
+      writer.setNote(&newMetaNote);
+      writer.writeToBuffer(&candidateElf);
```

Comment:
I don't think we should rewrite the ELF. It just adds more work to the compile process without good reasons.